File: training/passwordprotecting/python/storage.py
import guid
from user import User
from token import UserToken
from collections import defaultdict
"""
Storage simulates a generic storage device that could be a database
"""

class Storage:

  USERS:User = []
  USER_TOKENS:UserToken = []
  DATA = []

  expirationMsec = 10000

  def insertUser(username: str = None, password: str = None, publicName: str = None) -> int:
    if username == None or password == None:
      print("Failed to insert user, missing critical information")
      return False
    newUser = User(username, password, publicName)
    Storage.USERS.append(newUser)
    print(f"    .Storage.insertUser - Inserting: {newUser.id()} {username} {newUser._publicName}")
    return newUser.id()
  
  def doesUserExist(username:str = None):
    if username == None:
      return False
    for u in Storage.USERS:
      if u.username() == username:
        return True
    return False
  
  def getUser_RESTRICTED(username = None):
    if username == None:
      return None
    for u in Storage.USERS:
      if u.username() == username:
        return u
    print(f"    .Storage.getUserHash - Failed to find user: {username}")
    return None
  
  def loginUser(userId = -1):
    token = guid.generateGUID()
    print(f"    Generated token: {token}")
    Storage.USER_TOKENS.append(UserToken(userId, token))
    return token
  
  def isDuplicate(username:str = None):
    counter = 0
    for u in Storage.USERS:
      if u.username() == username:
        counter += 1
    return counter > 1
  
  #region TOKEN

  def isTokenValid(token:str = "0000"):
    found = None
    for uToken in Storage.USER_TOKENS:
      if uToken.token() == token:
        found = uToken
    if (found != None):
      print(f"    .Storage.isTokenValid::  expiration: ${Storage.expirationMsec}  token created: ${found._createdDateTime}")
      if (found._createdDateTime > Storage.expirationMsec):
        return found.getUserId()
      else:
        found._isValid = False
    return None

  def getTokenFromUserId(userId:int = -1):
    found = None
    for uToken in Storage.USER_TOKENS:
      if uToken.userId() == userId:
        found = uToken
    return found
  
  #endregion

  #region LOGOUT

  def logout(userId:int = -1):
    found = None
    for uToken in Storage.USER_TOKENS:
      if uToken.userId() == userId:
        found = uToken
    if found != None:
      found.invalidateToken()

  def logoutToken(token:str = None):
    found = None
    for uToken in Storage.USER_TOKENS:
      if uToken.token() == token:
        found = uToken
    if found != None:
      found.invalidateToken()

  def logoutUsername(username:str = None):
    usr = Storage.getUser_RESTRICTED(username)
    found = None
    for uToken in Storage.USER_TOKENS:
      if uToken.userId() == usr.UserId():
        found = uToken
    if found != None:
      found.invalidateToken()

  #endregion LOGOUT

  def shutdown (token: str = None):

    print('    .Storage.shutdown - Shutting down Storage')
    print('    .Storage - USERS content analysis:\n')
    if len(Storage.USERS) < 1:
      print("   - - - No users exist in storage - - -")
    for u in Storage.USERS:
      print(f"   USER:   {u.describe()}")

if __name__ == '__main__':

  print('\n ----- Storage class test code running ------- \n')

  testSto = Storage()
  print(f" - STORAGE - checkint for HAL : exists? {testSto.doesUserExist("HAL")}")

  user1name = "user1"
  user1pwd = "abcdef"
  print(f"\ninserting user: {user1name}")
  print(f"{testSto.insertUser(username= user1name, password= user1pwd, publicName="user_one")}")

  print(f"\nchecking for existing user: {user1name}")

  print(f"\n- STORAGE - USERS contains {user1name} : {testSto.doesUserExist(user1name)}")

  testSto.shutdown()
  

File: training/passwordprotecting/javascript/main/support/storage.js
import User from "./user.js"
import UserToken from "./token.js";
import generateGUID from "./guid.js";

const USERS = [];

const USER_TOKENS = [];

const DATA = [];

const expirationMsec = 10000;

class Storage {

  constructor() {
    // read users files to load array
    // 
  }

  // #region USER ACCESS

  /**
   * 
   * @param {string} username 
   * @param {string} password - GUID
   * @param {string} publicName 
   */
  insertUser(username, password, publicName) {
    let user = new User(username, password, publicName);
    USERS.push(user);
    console.log(`    .Storage.insertUser - Inserting: ${user.getId()} ${username} ${user._publicName}`)
    return user.getId();
  }

  /**
   * 
   * @param {string} username 
   * @returns {boolean}
   */
  doesUserExist(username) {
    return (USERS.find(user => user.getUsername() == username) != null);
  }

  /**
   * 
   * @param {string} username 
   * @returns {User}
   */
  getUser_RESTRICTED(username) {
    let index = USERS.findIndex(user => user.getUsername() == username);
    if (index > -1) {
      return USERS[index]
    }
    console.log(`    .Storage.getUserHash - Failed to find user: ${username}`)
    return null;
  }

  /**
   * Set the user to a logged in state by storing the token and user Id
   * @param {int} userId 
   * @returns {string} auth token 
   */
  loginUser(userId) {
    const token = generateGUID()[0]
    console.log(`    Generated token: ${token}`)
    USER_TOKENS.push(new UserToken(userId, token))
    return token;
  }



  
  /**
   * isTokenValid
   * @param {string} token - auth GUID 
   * @returns {int} user Id
   */
  isTokenValid(token) {
    currentCutoff = Date.now()
    const found = USER_TOKENS.find(ut => ut.getToken(expirationMsec) == token)
    console.log(`    .Storage.isTokenValid::  cutoff: ${currentCutoff}  token created: ${found?._createdDateTime}`);
    if (found) {
      if (found._createdDateTime + expirationMsec > currentCutoff) {
        return found?.getUserId();
      } else {
        found._isValid = false;
      }
    }
    return null;
  }

  /**
   * getTokenFromUserId
   * @param {int} userId 
   * @returns 
   */
  getTokenFromUserId(userId) {
    const found = USER_TOKENS.find(ut => ut.getUserId() == userId)
    return found?.getToken();
  }

  /**
   * logout
   * @param {int} userId 
   */
  logout(userId) {
    const found = USER_TOKENS.find(ut => ut.getUserId() == userId)
    if (found) {
      found.invalidateToken();
    }
  }

  /**
   * logoutUsername
   * @param {string} token GUID
   */
  logoutToken(token) {
    const found = USER_TOKENS.find(ut => ut.getToken() == token)
    if (found) {
      found.invalidateToken();
    }
  }

  /**
   * logoutUsername
   * @param {string} username 
   */
  logoutUsername(username) {
    const user = getUser_RESTRICTED(username);
    if (user) {
      const found = USER_TOKENS.find(ut => ut.getUserId() == user.getId())
      if (found) {
        found.invalidateToken();
      }
    }
  }

  /**
   * isDuplicate
   * @param {string} username 
   */
  isDuplicate(username) {
    let counter = 0;
    const nameCounter = (usr) => {
      if (usr._username == username) {
        counter++;
      }
    }
    USERS.forEach(nameCounter);
    console.log(`    .Storage.isDuplicate:: count of username ${username} is ${counter}`)
    if (counter == 0) console.log("    .Storage.isDuplicate:: username does not exist")
    return (counter > 1);
  }

  // #endregion USER ACCESS

  // #region STORAGE

  insertData(data, token, userId) {

  }

  getDataPermissions(dataId) {

  }

  readData(dataId) {

  }

  updateData(dataId, newData) {

  }

  deleteData(dataId) {

  }

  // #endregion STORAGE

  shutdown() {
    console.log(`    .Storage.shutdown - Shutting down storage`)

    // write to files

    const len = USERS.length;
    for (let i = 0; i < len; i++) {
      USERS[i].describe();
    }

  }
}

export default Storage;
